File: backend/src/models/Medication.js
import mongoose from 'mongoose';

export const MED_FORMS = Object.freeze([
  'tablet',
  'capsule',
  'insulin',
  'injection',
  'syrup',
  'inhaler',
  'topical',
  'other',
]);

/**
 * Whether the prescription for this medicine stands — the doctor's side.
 *
 * Only a clinician moves it, and only through services/medicationLifecycle.js:
 *
 *   active             in force
 *   completed          the course ran to its end date
 *   stopped_by_doctor  a clinician stopped it, or replaced the prescription
 *                      without carrying it over
 *   cancelled          the prescription it came from was voided — issued in
 *                      error, and the patient should never have been on it
 *   ended_legacy       ended before any of this was recorded. Who ended it is
 *                      not known, and the record does not pretend otherwise.
 */
export const PRESCRIPTION_STATE = Object.freeze({
  ACTIVE: 'active',
  COMPLETED: 'completed',
  STOPPED_BY_DOCTOR: 'stopped_by_doctor',
  CANCELLED: 'cancelled',
  ENDED_LEGACY: 'ended_legacy',
});

/**
 * Whether the patient is taking it — the patient's side.
 *
 * Stored: `taking` or `stopped_by_patient`. `not_started` is never stored; it
 * is what `taking` reads as before the start date (see takingStateOf).
 *
 * The two sides are separate fields on purpose. A patient pressing "Stop
 * taking" used to write the same `isActive: false` a doctor's stop wrote, so
 * the record could not say which had happened, and the patient's choice
 * rewrote the doctor's prescription.
 */
export const TAKING_STATE = Object.freeze({
  TAKING: 'taking',
  STOPPED_BY_PATIENT: 'stopped_by_patient',
  NOT_STARTED: 'not_started',
});

/**
 * A medication the patient is currently expected to take. Insulin is modelled
 * here too (form: 'insulin') so adherence and dose logging share one pipeline.
 *
 * ---- isActive ----------------------------------------------------------------
 *
 * Kept, and kept exact: prescription active AND patient taking. It is what every
 * reminder, today's schedule and every build of the app already read to decide
 * whether a dose is due, so a patient's stop and a doctor's stop both silence
 * the reminders without either of them being the other.
 */
const medicationSchema = new mongoose.Schema(
  {
    patient: { type: mongoose.Schema.Types.ObjectId, ref: 'User', required: true, index: true },

    name: { type: String, required: true, trim: true, maxlength: 160 },
    genericName: { type: String, trim: true, maxlength: 160 },
    form: { type: String, enum: MED_FORMS, default: 'tablet' },

    strength: { type: String, trim: true, maxlength: 60 }, // e.g. "500 mg", "100 IU/mL"
    dose: { type: String, trim: true, maxlength: 60 }, // e.g. "1 tablet", "12 units"

    // Local clock times "HH:mm" — reminders are scheduled on the device, so the
    // server stays timezone-agnostic.
    schedule: [
      {
        time: { type: String, required: true, match: [/^([01]\d|2[0-3]):[0-5]\d$/, 'time must be HH:mm'] },
        // The meal slot this dose is anchored to, so its time can be re-derived
        // when the patient changes their meal times. Absent for a manual time.
        slot: { type: String, enum: ['morning', 'noon', 'afternoon', 'night', 'bedtime'] },
        relationToMeal: {
          type: String,
          enum: ['before_meal', 'after_meal', 'with_meal', 'any'],
          default: 'any',
        },
      },
    ],
    daysOfWeek: {
      // 0 = Sunday. Empty means every day.
      type: [Number],
      default: [],
      validate: [(v) => v.every((d) => d >= 0 && d <= 6), 'daysOfWeek must be 0-6'],
    },

    // How the dose is given (PO/IV/SC/…). Presentation only; does not affect the
    // schedule.
    route: {
      type: String,
      enum: ['oral', 'iv', 'sc', 'im', 'topical', 'inhaled'],
      default: 'oral',
    },
    // PRN/SOS — taken only when required, so it carries an empty schedule and
    // arms no reminders.
    asNeeded: { type: Boolean, default: false },
    // Stat — a single immediate dose, not a recurring one.
    stat: { type: Boolean, default: false },
    // Every-other-day (EOD) etc.: 1 = daily, 2 = every other day. The device
    // scheduler only arms a dose on days matching this interval from startDate.
    dayInterval: { type: Number, default: 1, min: 1, max: 30 },

    startDate: { type: Date, default: Date.now },
    endDate: Date,
    isActive: { type: Boolean, default: true, index: true },
    // True once the patient overrides a reminder time by hand — a later
    // meal-time change then leaves this medicine's schedule untouched.
    timesCustomized: { type: Boolean, default: false },

    prescribedBy: { type: mongoose.Schema.Types.ObjectId, ref: 'User' },
    prescription: { type: mongoose.Schema.Types.ObjectId, ref: 'Prescription' },

    /// The practice whose prescription this is. Null for a medicine the patient
    /// added or photographed themselves, which is theirs and no practice's.
    practice: { type: mongoose.Schema.Types.ObjectId, ref: 'Practice', default: null, index: true },

    /// Earlier prescriptions this same medicine was continued from, oldest
    /// first. A renewal updates the row in place — its dose history stays on
    /// one medicine — and this keeps the trail the `prescription` field alone
    /// would overwrite.
    prescriptionHistory: [
      {
        _id: false,
        prescription: { type: mongoose.Schema.Types.ObjectId, ref: 'Prescription' },
        until: Date,
      },
    ],

    prescriptionState: {
      type: String,
      enum: Object.values(PRESCRIPTION_STATE),
      default: PRESCRIPTION_STATE.ACTIVE,
      index: true,
    },
    takingState: {
      type: String,
      enum: [TAKING_STATE.TAKING, TAKING_STATE.STOPPED_BY_PATIENT],
      default: TAKING_STATE.TAKING,
    },
    /// When the course ran out — its end date, recorded once it passed.
    completedAt: { type: Date, default: null },
    stoppedByDoctor: {
      at: Date,
      by: { type: mongoose.Schema.Types.ObjectId, ref: 'User' },
      reason: { type: String, trim: true, maxlength: 500 },
    },
    cancelled: {
      at: Date,
      by: { type: mongoose.Schema.Types.ObjectId, ref: 'User' },
      reason: { type: String, trim: true, maxlength: 500 },
    },
    /// Every time the patient stopped taking it, and when they started again.
    /// The patient's own history, in fields the doctor's side never reads as
    /// its own — nothing here changes what was prescribed.
    patientStops: [
      {
        _id: false,
        at: { type: Date, required: true },
        reason: { type: String, trim: true, maxlength: 300 },
        resumedAt: { type: Date, default: null },
      },
    ],

    /// How this medicine got here.
    ///
    /// `clinic` means a clinician issued it through the app, and prescribedBy
    /// names them. `scan` means the patient photographed a paper prescription —
    /// possibly one this clinic never wrote. The two were indistinguishable:
    /// a scanned medicine landed in the tracker with prescribedBy null, looking
    /// exactly like one of the doctor's own, and the doctor had no way to tell
    /// which of the medicines on the screen they were responsible for.
    source: { type: String, enum: ['clinic', 'scan', 'manual'], default: 'clinic', index: true },

    /// Who wrote the paper prescription, as read off the photograph.
    ///
    /// Unverified on purpose, and never used to decide anything. A doctor's
    /// name on a prescription is handwriting, a stamp or a letterhead, and OCR
    /// of it is not evidence — so it is carried as a label the reader can weigh
    /// and never as a gate. In particular it does NOT filter the medicine out:
    /// a list missing the steroid another doctor prescribed is more dangerous
    /// than a list with an unfamiliar name on it, because raised glucose with
    /// no visible cause is exactly what that omission produces.
    externalPrescriber: {
      name: { type: String, trim: true, maxlength: 160 },
      speciality: { type: String, trim: true, maxlength: 120 },
      clinic: { type: String, trim: true, maxlength: 160 },
      writtenOn: Date,
    },
    instructions: { type: String, maxlength: 600 },
  },
  { timestamps: true },
);

medicationSchema.index({ patient: 1, isActive: 1 });
// The completion sweep: standing prescriptions whose end date has passed.
medicationSchema.index({ prescriptionState: 1, endDate: 1 });

export const Medication = mongoose.model('Medication', medicationSchema);

File: mobile/lib/features/medications/domain/medication.dart
import 'med_shorthand.dart';

/// `Medication` object from API_CONTRACT.md §4.
class Medication {
  const Medication({
    required this.id,
    required this.name,
    required this.form,
    required this.strength,
    this.strengthExpected,
    this.strengthComposition,
    required this.dose,
    required this.schedule,
    required this.daysOfWeek,
    required this.isActive,
    this.genericName,
    this.startDate,
    this.endDate,
    this.instructions,
    this.route = 'oral',
    this.asNeeded = false,
    this.stat = false,
    this.dayInterval = 1,
    this.prescriptionState = 'active',
    this.takingState = 'taking',
    this.ownedBy = 'practice',
    this.completedAt,
    this.stoppedByDoctor,
    this.cancelled,
    this.stoppedTaking,
    this.alsoOnList = const [],
    this.changeableByYou,
  });

  /// The doctor's side: active | completed | stopped_by_doctor | cancelled |
  /// ended_legacy (ended before anybody recorded who ended it).
  final String prescriptionState;

  /// The patient's side: taking | stopped_by_patient | not_started.
  final String takingState;

  /// `practice` — a clinician prescribed it; `patient` — the patient added it.
  final String ownedBy;

  final DateTime? completedAt;
  final MedicineStop? stoppedByDoctor;
  final MedicineStop? cancelled;

  /// When and why the patient stopped taking it, while they have.
  final MedicineStop? stoppedTaking;

  /// Other prescriptions on this list for the same medicine — another strength,
  /// or another practice's. Never merged: which to keep is a clinical decision.
  final List<SameMedicine> alsoOnList;

  /// For a clinician: whether their practice may change this medicine. Null for
  /// the patient, who is told by [ownedBy] instead.
  final bool? changeableByYou;

  bool get prescriptionStands => prescriptionState == 'active';
  bool get stoppedByPatient => takingState == 'stopped_by_patient';
  bool get patientOwned => ownedBy == 'patient';

  final String id;
  final String name;
  final String? genericName;
  final String form;
  final String strength;

  /// What the clinic's brand list says this product's strength should be.
  /// Present only when it disagrees with [strength], or when none was recorded.
  /// Never applied automatically — a dose is changed by a person.
  final String? strengthExpected;

  /// "Metformin 500 mg + Glimepiride 1 mg", so the disagreement can be judged
  /// rather than merely reported.
  final String? strengthComposition;

  bool get hasStrengthMismatch => (strengthExpected ?? '').isNotEmpty;
  final String dose;
  final List<MedicationScheduleEntry> schedule;
  final List<String> daysOfWeek;
  final DateTime? startDate;
  final DateTime? endDate;
  final bool isActive;
  final String? instructions;

  /// oral | iv | sc | im | topical | inhaled.
  final String route;

  /// PRN/SOS — taken as needed, so it arms no reminders.
  final bool asNeeded;

  /// Stat — a single immediate dose.
  final bool stat;

  /// 1 = daily, 2 = every other day (EOD).
  final int dayInterval;

  factory Medication.fromJson(Map<String, dynamic> json) {
    return Medication(
      id: json['id']?.toString() ?? '',
      name: json['name']?.toString() ?? '',
      genericName: json['genericName'] as String?,
      form: json['form']?.toString() ?? '',
      strength: json['strength']?.toString() ?? '',
      strengthExpected: json['strengthExpected']?.toString(),
      strengthComposition: json['strengthComposition']?.toString(),
      dose: json['dose']?.toString() ?? '',
      schedule:
          (json['schedule'] as List<dynamic>? ?? const [])
              .map(
                (e) =>
                    MedicationScheduleEntry.fromJson(e as Map<String, dynamic>),
              )
              .toList(),
      daysOfWeek:
          (json['daysOfWeek'] as List<dynamic>? ?? const [])
              .map((e) => e.toString())
              .toList(),
      startDate:
          json['startDate'] == null
              ? null
              : DateTime.tryParse(json['startDate'].toString()),
      endDate:
          json['endDate'] == null
              ? null
              : DateTime.tryParse(json['endDate'].toString()),
      isActive: json['isActive'] as bool? ?? true,
      instructions: json['instructions'] as String?,
      route: json['route']?.toString() ?? 'oral',
      asNeeded: json['asNeeded'] as bool? ?? false,
      stat: json['stat'] as bool? ?? false,
      dayInterval: (json['dayInterval'] as num?)?.toInt() ?? 1,
      // A build older than the states reads every medicine it is sent as
      // active and taken, which is what `isActive` meant to it.
      prescriptionState: json['prescriptionState']?.toString() ?? 'active',
      takingState: json['takingState']?.toString() ?? 'taking',
      ownedBy: json['ownedBy']?.toString() ?? 'practice',
      completedAt: DateTime.tryParse(json['completedAt']?.toString() ?? '')?.toLocal(),
      stoppedByDoctor: MedicineStop.fromJson(json['stoppedByDoctor']),
      cancelled: MedicineStop.fromJson(json['cancelled']),
      stoppedTaking: MedicineStop.fromJson(json['stoppedTaking']),
      alsoOnList: [
        for (final o in (json['alsoOnList'] as List?) ?? const [])
          if (o is Map<String, dynamic>) SameMedicine.fromJson(o),
      ],
      changeableByYou: json['changeableByYou'] as bool?,
    );
  }
}

/// When a medicine was stopped, and the reason given.
class MedicineStop {
  const MedicineStop({required this.at, this.reason});

  final DateTime? at;
  final String? reason;

  static MedicineStop? fromJson(Object? json) {
    if (json is! Map<String, dynamic>) return null;
    return MedicineStop(
      at: DateTime.tryParse(json['at']?.toString() ?? '')?.toLocal(),
      reason: json['reason']?.toString(),
    );
  }
}

/// Another prescription on the list for the same medicine.
class SameMedicine {
  const SameMedicine({required this.id, this.strength, required this.samePractice});

  final String id;
  final String? strength;
  final bool samePractice;

  factory SameMedicine.fromJson(Map<String, dynamic> j) => SameMedicine(
    id: j['id']?.toString() ?? '',
    strength: j['strength']?.toString(),
    samePractice: j['samePractice'] as bool? ?? false,
  );
}

/// Result of `POST /medications/scan` — the medicines read from a prescription
/// photo and created in the tracker. [readable] is false when the photo could
/// not be read, so the UI can ask for a clearer one.
/// What the scan read, before any of it is saved.
///
/// [items] is a proposal, not a record. The server used to create the medicines
/// as it read them and hand back what it had already done — so a misread hour
/// was a live alarm before the patient could see it. It now reads, works out
/// the times, and stops; the patient looks, drops anything they do not
/// recognise, and only then is any of it written.
class PrescriptionScanResult {
  const PrescriptionScanResult({
    required this.readable,
    required this.created,
    this.items = const [],
    this.prescriber,
    this.note,
  });

  final bool readable;

  /// Empty from a scan. Filled by the confirm step with what was actually
  /// saved, so the sheet can say how many rather than guessing.
  final List<Medication> created;

  final List<ScannedMedicine> items;
  final Map<String, dynamic>? prescriber;
  final String? note;

  factory PrescriptionScanResult.fromJson(Map<String, dynamic> json) {
    return PrescriptionScanResult(
      readable: json['readable'] as bool? ?? false,
      created:
          (json['created'] as List<dynamic>? ?? const [])
              .map((e) => Medication.fromJson(e as Map<String, dynamic>))
              .toList(),
      items:
          (json['items'] as List<dynamic>? ?? const [])
              .whereType<Map<String, dynamic>>()
              .map(ScannedMedicine.fromJson)
              .toList(),
      prescriber: json['prescriber'] as Map<String, dynamic>?,
      note: json['note'] as String?,
    );
  }
}

/// One medicine the scan proposes, with the times it worked out.
class ScannedMedicine {
  const ScannedMedicine({
    required this.name,
    this.strength,
    this.dose,
    this.whenText,
    this.instructions,
    this.durationDays,
    this.schedule = const [],
  });

  final String name;
  final String? strength;
  final String? dose;
  /// The doctor's own words about when — "1 tab each AF Lunch". Shown beside
  /// the time so the patient checks a sentence against their paper rather than
  /// an hour they have no way to verify.
  final String? whenText;

  final String? instructions;
  final int? durationDays;

  /// 'HH:mm' times with their relation to a meal.
  final List<({String time, String relationToMeal})> schedule;

  factory ScannedMedicine.fromJson(Map<String, dynamic> j) => ScannedMedicine(
    name: j['name']?.toString() ?? '',
    strength: j['strength']?.toString(),
    dose: j['dose']?.toString(),
    whenText: j['whenText']?.toString(),
    instructions: j['instructions']?.toString(),
    durationDays: (j['durationDays'] as num?)?.toInt(),
    schedule: [
      for (final s in (j['schedule'] as List?) ?? const [])
        if (s is Map<String, dynamic> && s['time'] != null)
          (
            time: s['time'].toString(),
            relationToMeal: s['relationToMeal']?.toString() ?? 'any',
          ),
    ],
  );

  Map<String, dynamic> toJson() => {
    'name': name,
    if (strength != null) 'strength': strength,
    if (dose != null) 'dose': dose,
    if (instructions != null) 'instructions': instructions,
    if (durationDays != null) 'durationDays': durationDays,
    'schedule': [
      for (final s in schedule)
        {'time': s.time, 'relationToMeal': s.relationToMeal},
    ],
  };
}

/// The patient/dietician-facing plain-language dosing phrase, derived from the
/// medicine's shape (dose count, meal relation, route, and the PRN/Stat/EOD/HS
/// flags) via the shared shorthand dictionary — e.g. "Twice a day, after food".
extension MedicationDoseSummary on Medication {
  String get doseSummary => doseSummaryFrom(
    doseCount: schedule.length,
    relationToMeal: schedule.isNotEmpty ? schedule.first.relationToMeal : null,
    route: route,
    asNeeded: asNeeded,
    stat: stat,
    everyOtherDay: dayInterval > 1,
    atBedtime: schedule.any((s) => s.slot == 'bedtime'),
  );
}

class MedicationScheduleEntry {
  const MedicationScheduleEntry({
    required this.time,
    required this.relationToMeal,
    this.slot,
  });

  /// "HH:mm".
  final String time;

  /// before_meal | after_meal | with_meal | anytime.
  final String relationToMeal;

  /// morning | noon | afternoon | night | bedtime — so a bedtime (HS) dose can
  /// read "at bedtime" rather than a bare time.
  final String? slot;

  factory MedicationScheduleEntry.fromJson(Map<String, dynamic> json) {
    return MedicationScheduleEntry(
      time: json['time']?.toString() ?? '',
      relationToMeal: json['relationToMeal']?.toString() ?? 'anytime',
      slot: json['slot']?.toString(),
    );
  }
}

/// One row of `GET /medications/schedule/today`.
class MedicationScheduleSlot {
  const MedicationScheduleSlot({
    required this.medicationId,
    required this.name,
    required this.dose,
    required this.time,
    required this.relationToMeal,
    required this.status,
    this.logId,
    this.late = false,
    this.scheduledFor,
  });

  final String medicationId;
  final String name;
  final String dose;
  final String time;
  final String relationToMeal;

  /// The exact instant of this dose, as the server built it in the clinic's
  /// timezone. Sent back as-is when the dose is logged: rebuilt from the
  /// phone's clock, a phone in another timezone logged a time that is not
  /// one of the medicine's doses.
  final DateTime? scheduledFor;

  /// pending | taken | skipped | missed.
  final String status;
  final String? logId;

  /// Taken, but more than two hours after it was due.
  final bool late;

  factory MedicationScheduleSlot.fromJson(Map<String, dynamic> json) {
    return MedicationScheduleSlot(
      medicationId: json['medicationId']?.toString() ?? '',
      name: json['name']?.toString() ?? '',
      dose: json['dose']?.toString() ?? '',
      time: json['time']?.toString() ?? '',
      relationToMeal: json['relationToMeal']?.toString() ?? 'anytime',
      status: json['status']?.toString() ?? 'pending',
      logId: json['logId']?.toString(),
      late: json['late'] as bool? ?? false,
      scheduledFor: DateTime.tryParse(json['scheduledFor']?.toString() ?? ''),
    );
  }

  MedicationScheduleSlot copyWith({String? status, String? logId}) {
    return MedicationScheduleSlot(
      medicationId: medicationId,
      name: name,
      dose: dose,
      time: time,
      relationToMeal: relationToMeal,
      status: status ?? this.status,
      logId: logId ?? this.logId,
      late: late,
      scheduledFor: scheduledFor,
    );
  }
}

class TodaySchedule {
  const TodaySchedule({required this.date, required this.slots});

  final String date;
  final List<MedicationScheduleSlot> slots;

  factory TodaySchedule.fromJson(Map<String, dynamic> json) {
    return TodaySchedule(
      date: json['date']?.toString() ?? '',
      slots:
          (json['slots'] as List<dynamic>? ?? const [])
              .map(
                (e) =>
                    MedicationScheduleSlot.fromJson(e as Map<String, dynamic>),
              )
              .toList(),
    );
  }
}

/// One past dose in the patient's history (`GET /medications/schedule/history`).
class DoseHistoryEntry {
  const DoseHistoryEntry({
    required this.medicationId,
    required this.name,
    required this.time,
    required this.status,
    this.strength,
    this.form,
    this.relationToMeal,
    this.scheduledFor,
    this.takenAt,
    this.late = false,
  });

  final String medicationId;
  final String name;
  final String? strength;
  final String? form;
  final String time; // HH:mm (clinic wall-clock)
  final String? relationToMeal;
  final DateTime? scheduledFor;
  final DateTime? takenAt;

  /// Taken, but more than two hours after it was due.
  final bool late;

  /// taken | skipped | missed.
  final String status;

  factory DoseHistoryEntry.fromJson(Map<String, dynamic> j) => DoseHistoryEntry(
    medicationId: j['medicationId']?.toString() ?? '',
    name: j['name']?.toString() ?? '',
    strength: j['strength']?.toString(),
    form: j['form']?.toString(),
    time: j['time']?.toString() ?? '',
    relationToMeal: j['relationToMeal']?.toString(),
    scheduledFor:
        DateTime.tryParse(j['scheduledFor']?.toString() ?? '')?.toLocal(),
    takenAt: DateTime.tryParse(j['takenAt']?.toString() ?? '')?.toLocal(),
    status: j['status']?.toString() ?? 'missed',
    late: j['late'] as bool? ?? false,
  );
}

/// `GET /medications/adherence?days=30`.
class MedicationAdherence {
  const MedicationAdherence({
    required this.expected,
    required this.taken,
    required this.missed,
    required this.percentage,
    required this.perMedication,
  });

  final int expected;
  final int taken;
  final int missed;
  final num percentage;
  final List<PerMedicationAdherence> perMedication;

  factory MedicationAdherence.fromJson(Map<String, dynamic> json) {
    return MedicationAdherence(
      expected: (json['expected'] as num?)?.toInt() ?? 0,
      taken: (json['taken'] as num?)?.toInt() ?? 0,
      missed: (json['missed'] as num?)?.toInt() ?? 0,
      percentage: json['percentage'] as num? ?? 0,
      perMedication:
          (json['perMedication'] as List<dynamic>? ?? const [])
              .map(
                (e) =>
                    PerMedicationAdherence.fromJson(e as Map<String, dynamic>),
              )
              .toList(),
    );
  }
}

class PerMedicationAdherence {
  const PerMedicationAdherence({
    required this.medicationId,
    required this.name,
    required this.expected,
    required this.taken,
    required this.percentage,
  });

  final String medicationId;
  final String name;
  final int expected;
  final int taken;
  final num percentage;

  factory PerMedicationAdherence.fromJson(Map<String, dynamic> json) {
    return PerMedicationAdherence(
      medicationId: json['medicationId']?.toString() ?? '',
      name: json['name']?.toString() ?? '',
      expected: (json['expected'] as num?)?.toInt() ?? 0,
      taken: (json['taken'] as num?)?.toInt() ?? 0,
      percentage: json['percentage'] as num? ?? 0,
    );
  }
}
